src/analysis/processing/qgsalgorithmzonalhistogram.cpp

```diff
@@ -0,0 +1,323 @@
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "ZONES_FIELD" ), QObject::tr( "Group by field value" ),
+                QVariant(), QStringLiteral( "ZONES" ), QgsProcessingParameterField::Any, false, true ) );
+
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "PREFIX" ), QObject::tr( "Output column prefix" ), QStringLiteral( "HISTO_" ), false, true ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Output zones" ), QgsProcessing::TypeVectorPolygon ) );
+}
+
+QString QgsZonalHistogramAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm appends fields representing counts of each unique value from a raster layer contained within zones defined as polygons.\n\n"
+                      "Optionally, polygons can be regrouped using values from a provided field." );
+}
+
+QgsZonalHistogramAlgorithm *QgsZonalHistogramAlgorithm::createInstance() const
+{
+  return new QgsZonalHistogramAlgorithm();
+}
+
+bool QgsZonalHistogramAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsRasterLayer *layer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT" ), context );
+  mBand = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
+
+  if ( !layer )
+    return false;
+
+  mInterface.reset( layer->dataProvider()->clone() );
+  mHasNoDataValue = layer->dataProvider()->sourceHasNoDataValue( mBand );
+  mExtent = layer->extent();
+  mCrs = layer->crs();
+  mRasterUnitsPerPixelX = std::abs( layer->rasterUnitsPerPixelX() );
+  mRasterUnitsPerPixelY = std::abs( layer->rasterUnitsPerPixelX() );
+  mNbCellsXProvider = mInterface->xSize();
+  mNbCellsYProvider = mInterface->ySize();
+
+  return true;
+}
+
+QVariantMap QgsZonalHistogramAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  std::unique_ptr< QgsFeatureSource > zones( parameterAsSource( parameters, QStringLiteral( "ZONES" ), context ) );
+  if ( !zones )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "ZONES" ) ) );
+
+  QString zonesFieldName = parameterAsString( parameters, QStringLiteral( "ZONES_FIELD" ), context );
+  QgsAttributeList attributes;
+  int zonesFieldIndex = -1;
+  if ( !zonesFieldName.isEmpty() )
+  {
+    zonesFieldIndex = zones->fields().lookupField( zonesFieldName );
+    if ( zonesFieldIndex >= 0 )
+      attributes.append( zonesFieldIndex );
+  }
+
+  long count = zones->featureCount();
+  double step = count > 0 ? 100.0 / count : 1;
+  long current = 0;
+
+  QList< double > uniqueValues;
+  QMap< QVariant, QHash< double, qgssize > > featuresUniqueValues;
+
+  QgsFeatureRequest request; //.setDestinationCrs( sourceA.sourceCrs(), context.transformContext() );
```

Comment:
I'd suggest adding a comment here about the double- loop. E.g. how you have to iterate through zone features first to build up a list of all unique pixel values in order to determine the destination field list.